File: src/java/com/springbok/twobody/EquinoctialOrbit.java
package com.springbok.twobody;

import java.io.Serializable;

import com.springbok.center.SimulationConstants;

import Jama.Matrix;

/**
 * Describes an orbit using equinoctial elements by solving the generalize
 * Kepler's equation and computing position in the orbit plane and relative to
 * the Earth center.
 * 
 * Note the the implementation of the OrbitCorrection interface is a stub only.
 * 
 * @author Raymond LeClair
 */
@SuppressWarnings("serial")
public class EquinoctialOrbit implements Orbit, TwoBodyOrbit, EstimatedOrbit, Serializable {

	/** Direct or retrograde orbit indicator [-] */
	private int j;
	/** Semi-major axis [er] */
	private double a;
	/** y component of the eccentricity vector [-] */
	private double h;
	/** x component of the eccentricity vector [-] */
	private double k;
	/** y component of the nodal vector [-] */
	private double p;
	/** x component of the nodal vector [-] */
	private double q;
	/** Mean longitude [rad] */
	private double lambda;
	/** Epoch MJD calendar date */
	private ModJulianDate epoch;

	/** Method to sovle Kepler's equation: "newton" or "halley" */
	private String method;

	/** Mean motion [rad/s] */
	private double n;
	/** Orbital period [s] */
	private double T;

	@Override
	public ModJulianDate getEpoch() {
		return epoch;
	}

	/**
	 * Constructs an EquinoctialOrbit given equinoctial elements.
	 * 
	 * @param j
	 *            Direct or retrograde orbit indicator [-]
	 * @param a
	 *            Semi-major axis [er]
	 * @param h
	 *            Y component of the eccentricity vector [-]
	 * @param k
	 *            X component of the eccentricity vector [-]
	 * @param p
	 *            Y component of the nodal vector [-]
	 * @param q
	 *            X component of the nodal vector [-]
	 * @param lambda
	 *            Mean longitude [rad]
	 * @param epoch
	 *            Epoch MJD calendar date
	 * @param method
	 *            Method to sovle Kepler's equation: "newton" or "halley"
	 */
	public EquinoctialOrbit(int j, double a, double h, double k, double p, double q, double lambda, ModJulianDate epoch,
			String method) {

		// Fundamental values.

		this.j = j;
		this.a = a;
		this.h = h;
		this.k = k;
		this.p = p;
		this.q = q;
		this.lambda = lambda;
		this.epoch = epoch;

		this.method = method;

		// Derived values.

		this.n = meanMotion();
		this.T = orbitalPeriod();
	}

	/**
	 * Constructs an EquinoctialOrbit given equinoctial elements.
	 * 
	 * @param eqiOrb
	 *            An EquinoctialOrbit to copy
	 */
	public EquinoctialOrbit(EquinoctialOrbit eqiOrb) {

		// Fundamental values.

		this.j = eqiOrb.j;
		this.a = eqiOrb.a;
		this.h = eqiOrb.h;
		this.k = eqiOrb.k;
		this.p = eqiOrb.p;
		this.q = eqiOrb.q;
		this.lambda = eqiOrb.lambda;
		this.epoch = eqiOrb.epoch;

		this.method = eqiOrb.method;

		// Derived values.

		this.n = meanMotion();
		this.T = orbitalPeriod();
	}

	/**
	 * Constructs an EquinoctialOrbit.
	 */
	public EquinoctialOrbit() {

		// Fundamental values.

		this.j = 0;
		this.a = 0.0;
		this.h = 0.0;
		this.k = 0.0;
		this.p = 0.0;
		this.q = 0.0;
		this.lambda = 0.0;
		this.epoch = new ModJulianDate(0.0);

		this.method = "halley";

		// Derived values.

		this.n = 0.0;
		this.T = 0.0;
	}

	/**
	 * Sets method for solving Kepler's equation.
	 * 
	 * @param method
	 *            Method for solving Kepler's equation: "newton" or "halley"
	 */
	public void set_method(String method) {
		if (!method.equals("newton") && !method.equals("halley")) {
			throw new IllegalArgumentException("Method must be either \"newton\" or \"halley\".");
		}
		this.method = method;
	}

	/**
	 * Gets method for solving Kepler's equation.
	 * 
	 * @return Method for solving Kepler's equation
	 */
	public String get_method() {
		return method;
	}

	/**
	 * Computes mean motion (MG-2.35).
	 * 
	 * @return Mean motion [rad/s]
	 */
	public double meanMotion() {
		return Math.sqrt(EarthConstants.GM_oplus / Math.pow(EarthConstants.R_oplus * a, 3));
	}

	/**
	 * Computes the mean longitude at the given date (LCVF-3-271).
	 * 
	 * @param dNm
	 *            MJD calendar date at which mean longitude occurs
	 * @return Mean longitude [rad]
	 */
	public double meanPosition(ModJulianDate dNm) {
		return Coordinates.checkWrap(lambda + n * (dNm.getOffset(epoch)));
		// [rad] = [rad] + [rad/s] * [ms] / [ms/s]
	}

	/**
	 * Computes orbital period.
	 * 
	 * @return Orbital period [s]
	 */
	public double orbitalPeriod() {
		return (2 * Math.PI) * Math.sqrt(Math.pow(EarthConstants.R_oplus * a, 3) / EarthConstants.GM_oplus);
		// [s/rev] = [rad/rev] * sqrt(([km] * [-])^3 / [km^3/s^2])
	}

	/**
	 * Solve generalized Kepler's equation using Newton's or Halley's method to
	 * solve for the eccentric longitude (LCVF-3-230).
	 * 
	 * @param lambda
	 *            mean longitude [rad]
	 * @return Eccentric longitude [rad]
	 */
	public double keplersEquation(double lambda) {
		double F_i, F_iplus1;
		double f_i, f_p_i, f_pp_i;
		int nItn = 0;

		F_i = lambda;

		f_i = F_i + h * Math.cos(F_i) - k * Math.sin(F_i) - lambda;
		f_p_i = 1 - h * Math.sin(F_i) - k * Math.cos(F_i);

		if (method.equals("newton")) {
			F_iplus1 = F_i - F_i / f_p_i;
		} else if (method.equals("halley")) {
			f_pp_i = -h * Math.cos(F_i) + k * Math.sin(F_i);

			F_iplus1 = F_i - (2 * f_i * f_p_i) / (2 * Math.pow(f_p_i, 2) - F_i * f_pp_i);
		} else {
			throw new IllegalStateException("Invalid method.");
		}
		nItn = 0;
		while (Math.abs(F_iplus1 - F_i) > SimulationConstants.precision_E) {
			nItn++;
			if (nItn > SimulationConstants.max_iteration) {
				System.out.println("Maximum iterations exceeded.");
				break;
			}
			F_i = F_iplus1;

			f_i = F_i + h * Math.cos(F_i) - k * Math.sin(F_i) - lambda;
			f_p_i = 1 - h * Math.sin(F_i) - k * Math.cos(F_i);

			if (method.equals("newton")) {
				F_iplus1 = F_i - F_i / f_p_i;
			} else if (method.equals("halley")) {
				f_pp_i = -h * Math.cos(F_i) + k * Math.sin(F_i);

				F_iplus1 = F_i - (2 * f_i * f_p_i) / (2 * Math.pow(f_p_i, 2) - F_i * f_pp_i);
			}
		}
		return F_iplus1;
	}

	/**
	 * Computes orbital plane inertial position vector (LCVF-3-236).
	 * 
	 * @param F
	 *            Eccentric longitude [rad]
	 * @return Orbital plane inertial position vector [er]
	 */
	public Matrix r_goi(double F) {
		double beta;
		double X_1, Y_1;
		beta = 1 / (1 + Math.sqrt(1 - Math.pow(h, 2) - Math.pow(k, 2)));
		X_1 = a * ((1 - Math.pow(h, 2) * beta) * Math.cos(F) + h * k * beta * Math.sin(F) - k);
		Y_1 = a * ((1 - Math.pow(k, 2) * beta) * Math.sin(F) + h * k * beta * Math.cos(F) - h);
		double[][] elements = { { X_1 }, { Y_1 }, { 0.0 } };
		Matrix matrix = new Matrix(elements);
		// System.out.println("r_goi"); matrix.print(16, 8);
		return matrix;
	}

	/**
	 * Computes geocentric equatorial intertial position vector (LCVF-3-238).
	 * 
	 * @param dNm
	 *            MJD calendar date at which the position vector occurs
	 * @return Geocentric equatorial intertial position vector [er]
	 */
	public Matrix r_gei(ModJulianDate dNm) {
		double m = 1 + Math.pow(p, 2) + Math.pow(q, 2);
		double[][] f_elements = { { (1 - Math.pow(p, 2) + Math.pow(q, 2)) / m }, { (2 * p * q) / m },
				{ (-2 * p * j) / m } };
		Matrix f = new Matrix(f_elements);
		double[][] g_elements = { { (2 * p * q * j) / m }, { ((1 + Math.pow(p, 2) - Math.pow(q, 2)) * j) / m },
				{ (2 * q) / m } };
		Matrix g = new Matrix(g_elements);
		Matrix r_goi = r_goi(keplersEquation(meanPosition(dNm)));
		double X_1 = r_goi.get(0, 0);
		double Y_1 = r_goi.get(1, 0);
		Matrix r_gei = f.times(X_1).plus(g.times(Y_1));
		// System.out.println("r_gei"); r_gei.print(16, 8);
		return r_gei;
	}

	/**
	 * Gets direct or retrograde orbit indicator [-]
	 * 
	 * @return Direct or retrograde orbit indicator [-]
	 */
	public int get_j() {
		return j;
	}

	/**
	 * Sets semi-major axis [er].
	 * 
	 * @param a
	 *            Semi-major axis [er]
	 */
	public void set_a(double a) {
		this.a = a;
		n = meanMotion();
		T = orbitalPeriod();
	}

	/**
	 * Gets semi-major axis [er].
	 * 
	 * @return Semi-major axis [er]
	 */
	public double get_a() {
		return a;
	}

	/**
	 * Sets y component of the eccentricity vector [-].
	 * 
	 * @param h
	 *            Y component of the eccentricity vector [-]
	 */
	public void set_h(double h) {
		this.h = h;
	}

	/**
	 * Gets y component of the eccentricity vector [-].
	 * 
	 * @return Y component of the eccentricity vector [-]
	 */
	public double get_h() {
		return h;
	}

	/**
	 * Sets x component of the eccentricity [-].
	 * 
	 * @param k
	 *            X component of the eccentricity [-]
	 */
	public void set_k(double k) {
		this.k = k;
	}

	/**
	 * Gets x component of the eccentricity [-].
	 * 
	 * @return X component of the eccentricity [-]
	 */
	public double get_k() {
		return k;
	}

	/**
	 * Sets y component of the nodal vector [-].
	 * 
	 * @param p
	 *            Y component of the nodal vector [-]
	 */
	public void set_p(double p) {
		this.p = p;
	}

	/**
	 * Gets y component of the nodal vector [-].
	 * 
	 * @return Y component of the nodal vector [-]
	 */
	public double get_p() {
		return p;
	}

	/**
	 * Sets x component of the nodal vector [-].
	 * 
	 * @param q
	 *            X component of the nodal vector [-]
	 */
	public void set_q(double q) {
		this.q = q;
	}

	/**
	 * Gets x component of the nodal vector [-].
	 * 
	 * @return X component of the nodal vector [-]
	 */
	public double get_q() {
		return q;
	}

	/**
	 * Sets mean longitude [rad].
	 * 
	 * @param lambda
	 *            Mean longitude [rad]
	 */
	public void set_lambda(double lambda) {
		this.lambda = lambda;
	}

	/**
	 * Gets mean longitude [rad].
	 * 
	 * @return Mean longitude [rad]
	 */
	public double get_lambda() {
		return lambda;
	}

	/**
	 * Sets epoch MJD calendar date.
	 * 
	 * @param epoch
	 *            MJD calendar date
	 */
	public void set_epoch(ModJulianDate epoch) {
		this.epoch = epoch;
	}

	/**
	 * Gets epoch MJD calendar date.
	 * 
	 * @return Epoch MJD calendar date
	 */
	public ModJulianDate get_epoch() {
		return epoch;
	}

	/**
	 * Sets eccentricity [-].
	 * 
	 * @param e
	 *            Eccentricity [-]
	 */
	public void set_e(double e) {
		;
	}

	/**
	 * Gets eccentricity [-].
	 * 
	 * @return Eccentricity [-]
	 */
	public double get_e() {
		return 0.0;
	}

	/**
	 * Sets inclination [rad].
	 * 
	 * @param i
	 *            Inclination [rad]
	 */
	public void set_i(double i) {
		;
	}

	/**
	 * Gets inclination [rad].
	 * 
	 * @return Inclination [rad]
	 */
	public double get_i() {
		return 0.0;
	}

	/**
	 * Sets right ascension of the ascending node [rad].
	 * 
	 * @param Omega
	 *            Right ascension of the ascending node [rad]
	 */
	public void set_Omega(double Omega) {
		;
	}

	/**
	 * Gets right ascension of the ascending node [rad].
	 * 
	 * @return Right ascension of the ascending node [rad]
	 */
	public double get_Omega() {
		return 0.0;
	}

	/**
	 * Sets argument of perigee [rad].
	 * 
	 * @param omega
	 *            Argument of perigee [rad]
	 */
	public void set_omega(double omega) {
		;
	}

	/**
	 * Gets argument of perigee [rad].
	 * 
	 * @return Argument of perigee [rad]
	 */
	public double get_omega() {
		return 0.0;
	}

	/**
	 * Sets mean anomaly [rad].
	 * 
	 * @param M
	 *            Mean anomaly [rad]
	 */
	public void set_M(double M) {
		;
	}

	/**
	 * Gets mean anomaly [rad].
	 * 
	 * @return Mean anomaly [rad]
	 */
	public double get_M() {
		return 0.0;
	}

	/**
	 * Gets mean motion [rad/s].
	 * 
	 * @return Mean motion [rad/s]
	 */
	public double get_n() {
		return n;
	}

	/**
	 * Gets orbital period [s].
	 * 
	 * @return Orbital period [s]
	 */
	public double get_T() {
		return T;
	}

	@Override
	public Orbit copy() {
		return new EquinoctialOrbit(this);
	}
}

File: src/matlab/com/springbok/twobody/EquinoctialOrbit.m
classdef EquinoctialOrbit < Orbit & TwoBodyOrbit
% Describes an orbit using equinoctial elements by solving the generalized
% Kepler's equation and computing position in the orbit plane and
% relative to the Earth center.
  
% Copyright (C) 2016 Springbok LLC <http://springbok.io/> All rights reserved.

  properties (SetAccess = private, GetAccess = public)
    
    % Direct or retrograde orbit indicator [-]
    j
    % Semi-major axis [er]
    a
    % Y component of the eccentricity vector [-]
    h
    % X component of the eccentricity  [-]
    k
    % Y component of the nodal vector [-]
    p
    % X component of the nodal vector [-]
    q
    % Mean longitude [rad]
    lambda
    % Epoch date number
    epoch
    
    % Method to solve Kepler's equation: 'newton' or 'halley'
    method = 'halley';
    
    % Mean motion [rad/s]
    n
    % Orbital period [s]
    T
    
  end % properties (SetAccess = private, GetAccess = public)
  
  methods
    
    function this = EquinoctialOrbit(j, a, h, k, p, q, lambda, epoch, method)
    % Constructs an EquinoctialOribt given equinoctial elements.
    %
    % Parameters
    %   j - Direct or retrograde orbit indicator [-]
    %   a - Semi-major axis [er]
    %   h - Y component of the eccentricity vector [-]
    %   k - X component of the eccentricity vector [-]
    %   p - Y component of the nodal vector [-]
    %   q - X component of the nodal vector [-]
    %   lambda - Mean longitude [rad]
    %   epoch - Epoch date number
    %   method - Method to solve Kepler's equation: 'newton' or 'halley'
      
    % No argument constructor
      if nargin == 0
        return;
        
      end % if
      
      % Assign properties. Set numeric properties directly since
      % setters have side effects.
      this.j      = TypeUtility.set_numeric(j);
      this.a      = TypeUtility.set_numeric(a);
      this.h      = TypeUtility.set_numeric(h);
      this.k      = TypeUtility.set_numeric(k);
      this.p      = TypeUtility.set_numeric(p);
      this.q      = TypeUtility.set_numeric(q);
      this.lambda = TypeUtility.set_numeric(lambda);
      this.epoch  = TypeUtility.set_numeric(epoch);

      this.set_method(method);

      % Compute derived properties
      this.n = meanMotion(this);
      this.T = orbitalPeriod(this);
      
    end % EquinoctialOrbit()
    
    function that = copy(this)
    % Copies an EquinoctialOribt given equinoctial elements.
    %
    % Returns
    %   that - A new EquinoctialOrbit instance
      that = EquinoctialOrbit(this.j, this.a, this.h, this.k, this.p, ...
                              this.q, this.lambda, this.epoch, ...
                              this.method);

    end % copy()

    function n = meanMotion(this)
    % Computes mean motion (MG-2.35).
    %
    % Returns
    %   n - Mean motion [rad/s]
      n = sqrt(EarthConstants.GM_oplus ...
               ./ (EarthConstants.R_oplus * this.a).^3);
      % [rad/s] = [ [km^3/s^2] / [ [km/er] * [er] ]^3 ]^(1/2)
      
    end % meanMotion()
    
    function lambda = meanPosition(this, dNm)
    % Computes the mean longitude at the given date (LCVF-3-271).
    %
    % Parameters
    %   dNm - Date number at which mean longitude occurs
    %
    % Returns
    %   lambda - Mean longitude [rad]
      lambda = Coordinates.check_wrap(this.lambda ...
                                      + this.n * (dNm - this.epoch) * 86400);
      % [rad] = [rad] + [rad/s] * [decimal day] * [s/day]
      
    end % meanPosition()
    
    function T = orbitalPeriod(this)
    % Computes orbital period.
    %
    % Returns
    %   T - Orbital period [s]
      T = (2 * pi) * sqrt((EarthConstants.R_oplus * this.a).^3 ...
                          / EarthConstants.GM_oplus);
      % [s/rev] = [rad/rev] * sqrt(([km] * [-])^3 / [km^3/s^2])
      
    end % orbitalPeriod()
    
    function F = keplersEquation(this, lambda)
    % Solve generalized Kepler's equation using Newton's or Halley's
    % method to solve for the eccentric longitude (LCVF-3-230).
    %
    % Parameters
    %   lambda - Mean longitude [rad]
    %
    % Returns
    %   F - Eccentric longitude, radians
      
      F_i = lambda;
      
      f_i   = F_i + this.h * cos(F_i) - this.k * sin(F_i) - lambda;
      f_p_i = 1 - this.h * sin(F_i) - this.k * cos(F_i);
      
      switch this.method
        case 'newton'
          F_iplus1 = F_i - F_i / f_p_i;
          
        case 'halley'
          f_pp_i = - this.h * cos(F_i) + this.k * sin(F_i);
          
          F_iplus1 = F_i - (2 * f_i * f_p_i) / (2 * f_p_i^2 - F_i * f_pp_i);
          
        otherwise
          MEx = MException('Springbok:IllegalArgumentException', 'Invalid method.');
          throw(MEx);
          
      end % switch
      
      nItn = 0;
      while abs(F_iplus1 - F_i) > this.precision_E
        nItn = nItn + 1;
        if nItn > this.max_iteration
          Simlogger.printMsg(Simlogger.INFO, 'Maximum iterations exceeded.');
          break;
          
        end % if
        
        F_i = F_iplus1;
        
        f_i   = F_i + this.h * cos(F_i) - this.k * sin(F_i) - lambda;
        f_p_i = 1 - this.h * sin(F_i) - this.k * cos(F_i);
        
        switch this.method
          case 'newton'
            F_iplus1 = F_i - f_i / f_p_i;
            
          case 'halley'
            f_pp_i = - this.h * cos(F_i) + this.k * sin(F_i);
            
            F_iplus1 = F_i - (2 * f_i * f_p_i) / (2 * f_p_i^2 - f_i * f_pp_i);
            
        end % switch
        
      end % while abs(F_iplus1 - F_i) > this.precision_E
      
      F = F_iplus1;
      
    end % keplersEquation()
    
    function r_goi = r_goi(this, F)
    % Computes orbital plane inertial position vector (LCVF-3-236).
    %
    % Parameters
    %   F - Eccentric longitude [rad]
    %
    % Returns
    %   r_goi - Orbital plane inertial position vector [er]
      
      beta = 1 / (1 + sqrt(1 - this.h^2 - this.k^2));
      
      X_1 = this.a * ((1 - this.h^2 * beta) * cos(F) ...
                      + this.h * this.k * beta * sin(F) - this.k);
      Y_1 = this.a * ((1 - this.k^2 * beta) * sin(F) ...
                      + this.h * this.k * beta * cos(F) - this.h);
      
      r_goi = [X_1; ...
               Y_1; ...
               0];
      
    end % r_goi()
    
    function r_gei = r_gei(this, dNm)
    % Computes geocentric equatorial inertial position vector
    % (LCVF-3-238).
    %
    % Parameters
    %   dNm - Date number at which the position vector occurs
    %
    % Returns
    %   r_gei - Geocentric equatorial inertial position vector [er]
      
      f = [1 - this.p^2 + this.q^2; ...
           2 * this.p * this.q; ...
           - 2 * this.p * this.j] / (1 + this.p^2 + this.q^2);
      
      g = [2 * this.p * this.q * this.j; ...
           (1 + this.p^2 - this.q^2) * this.j; ...
           + 2 * this.q] / (1 + this.p^2 + this.q^2);
      
      r_goi = this.r_goi(this.keplersEquation(this.meanPosition(dNm)));
      r_gei = r_goi(1) * f + r_goi(2) * g; 
      
    end % r_gei()
    
    function set_a(this, a)
    % Sets semi-major axis [er].
    %
    % Parameters
    %   a - Semi-major axis [er]
      this.a = a;
      this.n = meanMotion(this);
      this.T = orbitalPeriod(this);
      
    end % set_a()
    
    function set_h(this, h)
    % Sets y component of the eccentricity vector [-].
    %
    % Parameters
    %   h - Y component of the eccentricity vector [-]
      this.h = h;
      
    end % set_h()
    
    function set_k(this, k)
    % Sets x component of the eccentricity  [-].
    %
    % Parameters
    %   k - X component of the eccentricity  [-]
      this.k = k;
      
    end % set_k()
    
    function set_p(this, p)
    % Sets y component of the nodal vector [-].
    %
    % Parameters
    %   p - Y component of the nodal vector [-]
      this.p = p;
      
    end % set_p()
    
    function set_q(this, q)
    % Sets x component of the nodal vector [-].
    %
    % Parameters
    %   q - X component of the nodal vector [-]
      this.q = q;
      
    end % set_q()
    
    function set_lambda(this, lambda)
    % Sets mean longitude [rad].
    %
    % Parameters
    %   lambda - Mean longitude [rad]
      this.lambda = lambda;
      
    end % set_lambda()
    
    function set_epoch(this, epoch)
    % Sets epoch date number.
    % Parameters
    %   epoch - Date number
    %
      this.epoch = epoch;
      
    end % set_epoch()
    
    function set_method(this, method)
    % Sets method for solving Kepler's equation.
    %
    % Parameters
    %   method - Method for solving Kepler's equation: 'newton' or
    %         'halley'
      if ~isequal(method, 'newton') && ~isequal(method, 'halley')
        MEx = MException('Springbok:IllegalArgumentException', ...
                         'Method must be either newton or halley.');
        throw(MEx);
        
      end % if
      this.method = method;
      
    end % set_method()
    
    function is = isEmpty(this)
    % Determines if orbit properties are empty, or not.
      is = isempty(this.j) ...
           && isempty(this.a) ...
           && isempty(this.h) ...
           && isempty(this.k) ...
           && isempty(this.p) ...
           && isempty(this.q) ...
           && isempty(this.lambda) ...
           && isempty(this.epoch) ...
           && isempty(this.n) ...
           && isempty(this.T);
      % && isempty(this.method) ...
      
    end % isEmpty()
    
    function is = isEqual(this, that)
    % Determines if two orbits are equal, or not.
      
      if this.isEmpty() || that.isEmpty()
        is = this.isEmpty() && that.isEmpty();

      else
        t = [];
        t = [t; isequal(this.j, that.j)];
        t = [t; isequal(this.a, that.a)];
        t = [t; isequal(this.h, that.h)];
        t = [t; isequal(this.k, that.k)];
        t = [t; isequal(this.p, that.p)];
        t = [t; isequal(this.q, that.q)];
        t = [t; isequal(this.lambda, that.lambda)];
        t = [t; isequal(this.epoch, that.epoch)];
        t = [t; isequal(this.method, that.method)];
        t = [t; abs(this.n - that.n) < TestUtility.HIGH_PRECISION];
        t = [t; abs(this.T - that.T) < TestUtility.HIGH_PRECISION];

        is = min(t);
      
      end % if
      
    end % isEqual()

  end % methods
  
end % classdef
